server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2178,16 +2220,19 @@ private void innerOpenEngineAndTranslog(LongSupplier globalCheckpointSupplier) t
                 + recoveryState.getRecoverySource()
                 + "] but got "
                 + getRetentionLeases();
-        synchronized (engineMutex) {
-            assert currentEngineReference.get() == null : "engine is running";
+        engineLock.writeLock().lock();
+        try {
+            assert currentEngine == null : "engine is running";
             verifyNotClosed();
             // we must create a new engine under mutex (see IndexShard#snapshotStoreMetadata).
             final Engine newEngine = createEngine(config);
```

Comment:
I wonder if we should call this before acquiring the write lock. I'm thinking about cases where the engine might take some time to initialize (i.e. in stateless the network might be slow and we could potentially block the cluster state applier thread if the shard is closed in between).

Reply:
I think this is unmodified by this PR? We can look at it as a follow-up then.

Reply:
This is indeed unchanged. I have not seen this (in production monitoring) to be an issue that needs to be addressed immediately though.

Reply:
The `InternalEngine` constructor does writes in that it trims unsafe commits. I suppose we'd be protected by the shard lock wrt a reallocation to the same node. But we'd need some care if we were to do this - hence a strong preference for deferring that work (out of this PR at least).

Reply:
That's a good point.

> But we'd need some care if we were to do this - hence a strong preference for deferring that work (out of this PR at least).

I wonder if we should do this at all, at least until we're sure it is a source of issues? Because I think that creating/closing engines while holding the write lock will help later for dynamic hollowing, where we could have a succession of hollow->unhollow->hollow. If we were creating or closing engines outside of the write lock, we could end up having two internal engines trying to acquire the index writer lock I think.